pkg/operator/storageclasscontroller/vmware.go

```diff
@@ -89,6 +89,58 @@ func (v *storagePolicyAPI) getDefaultDatastore(ctx context.Context) (*mo.Datasto
 	return &dsMo, nil
 }
 
+func (v *storagePolicyAPI) getDatastore(ctx context.Context, dcName string, dsName string) (*mo.Datastore, error) {
+	vmClient := v.vcenterApiConnection.Client
+	finder := find.NewFinder(vmClient.Client, false)
+
+	dc, err := finder.Datacenter(ctx, dcName)
+	if err != nil {
+		return nil, fmt.Errorf("failed to access datacenter %s: %s", dcName, err)
+	}
+
+	finder.SetDatacenter(dc)
+	ds, err := finder.Datastore(ctx, dsName)
+	if err != nil {
+		return nil, fmt.Errorf("failed to access datastore %s: %s", dsName, err)
+	}
+
+	var dsMo mo.Datastore
+	pc := property.DefaultCollector(dc.Client())
+	properties := []string{DatastoreInfoProperty, SummaryProperty}
+	err = pc.RetrieveOne(ctx, ds.Reference(), properties, &dsMo)
+	if err != nil {
+		return nil, fmt.Errorf("error getting properties of datastore %s: %v", dsName, err)
+	}
+	return &dsMo, nil
+}
+
+func (v *storagePolicyAPI) createZonalStoragePolicy(ctx context.Context) (string, error) {
+	var err error
+	failureDomains := v.infra.Spec.PlatformSpec.VSphere.FailureDomains
+	if len(failureDomains) <= 1 {
+		return "", fmt.Errorf("cluster has only one failure domain defined")
+	}
+
+	for _, failureDomain := range failureDomains {
+		dataCenter := failureDomain.Topology.Datacenter
+		dataStore := failureDomain.Topology.Datastore
+		ds, err := v.getDatastore(ctx, dataCenter, dataStore)
+		if err != nil {
+			return v.policyName, fmt.Errorf("unable to fetch datastore %s: %v", dataStore, err)
+		}
+		err = v.createOrUpdateTag(ctx, ds)
+		if err != nil {
+			return v.policyName, fmt.Errorf("error creating or updating tag %s: %v", v.tagName, err)
+		}
```

Comment:
Instead of stopping the execution, should it try the next one?

Reply:
I think it might be better to hard error out, rather than continuing with tagging whatever datastores we could access, because that could result in a storagepolicy which has unpredictable behaviour.

Reply:
My point was when you have say 5 failureDomains/datastores, then you successfully tag 3 of them and the 4th fail. Sounds like it's better to either:
1. Try tagging all of them and return an aggregated error
2. Return early in case of error, but don't leave behind tagged datastores

Reply:
okay I have aggregated the errors and returned them. I think that is a good idea in case we have multiple datastores and we don't have access to them, so as rather than returning one error at a time - we will return error about all inaccessible datastores at the same time. 

I have pushed that change to - https://github.com/openshift/vmware-vsphere-csi-driver-operator/pull/125 however, which also includes #121